src/main/java/com/bwarelabs/BigTableToCosWriter.java

```diff
@@ -443,4 +508,44 @@ public List<String[]> splitRangeTx() {
         }
         return intervals;
     }
+
+    private CompletableFuture<AsyncConnection> getConn() {
+        CompletableFuture<AsyncConnection> f = future.get();
+        if (f != null) {
+            return f;
+        }
+        for (;;) {
+            if (future.compareAndSet(null, new CompletableFuture<>())) {
+                CompletableFuture<AsyncConnection> toComplete = future.get();
+                addListener(ConnectionFactory.createAsyncConnection(this.configuration), (conn, error) -> {
+                    if (error != null) {
+                        toComplete.completeExceptionally(error);
+                        // reset the future holder so we will get a chance to recreate an async
+                        // connection at next try.
+                        future.set(null);
+                        return;
+                    }
+                    toComplete.complete(conn);
+                });
+                return toComplete;
+            } else {
+                f = future.get();
+                if (f != null) {
+                    return f;
+                }
+            }
+        }
+    }
+
+    private static class BatchCounter {
```

Comment:
this was added due to `Variable used in lambda expression should be final or effectively final` but it can t be final since i need to increment